### PR TITLE
MAP-44-45 - HOURS VIEWS DRILLDOWNS

### PR DESCRIPTION
**Pull request by calderonmluis**
Drill-downs for hours details

### Summary

This PR contains:

- Drill-downs for MAP-44/45 (hours drill-downs)
- Updates to test assertions to help track down exact errors

### Test Plan

- [x] Validate test coverage and success
- [x] Validate loading Dashboards and functionality

### Checklist

- [x] Passes PEP8 Linting
- [-] Documentation: <provide link>

### Screenshots

Screen Shot 2019-12-26 at 4 56 00 PM
Screen Shot 2019-12-26 at 4 52 14 PM

### CODE REVIEW

File: generator/vogtle_data_generator.py
Comment (TimothyHelton) [inline generator/vogtle_data_generator.py]: - [x] remove `rows` from docstring
- [x] add `itaac_id` docstring
- [x] add `effort_id` docstring
- [x] add `effort_range` docstring

File: generator/vogtle_data_generator.py
Comment (TimothyHelton) [inline generator/vogtle_data_generator.py]: - [ ] `seperated` should be `separated`

```suggestion
        """Generate string, bar separated value of itaac_effort entry
```

File: generator/vogtle_data_generator.py
Comment (TimothyHelton) [inline generator/vogtle_data_generator.py]: What data type are you expecting the user to supply for `effort_range`?

If it's not a list then consider converting `effort_range` only once.

```python
effort_range = list(effort_range)
# If the supplied data type could be a str or list for example then do the following:
# effort_range = [str] if isinstance(effort_range, str) else effort_range
actual = random.choice(effort_range)
estimate = random.choice(effort_range)
```

File: generator/vogtle_data_generator.py
Comment (TimothyHelton) [inline generator/vogtle_data_generator.py]: TODO on future PR:
Convert string statements to `f strings`.  Only use old-style string formatting for logging statements, since the logging module is optimized for the old-style.

Example f string:
```python
        return f'\n{effort_id}|{itaac_id}|{effort_type}|{actual}|{estimate}'
```

File: generator/vogtle_data_generator.py
Comment (TimothyHelton) [inline generator/vogtle_data_generator.py]: - [x] Just return he previous statement in lieu of creating a variable to return.

File: generator/vogtle_data_generator.py
Comment (TimothyHelton) [inline generator/vogtle_data_generator.py:84]: - [ ] Use the `randint` function to select a random integer (e.g. `actual = random.randint(0, 120)`)
- [x] Keep the `output_file.write` lines together, so there is only one place to look.
- [ ] Move the `output.append` statement into the if clause to replace `actual` and `estimate`.
- [ ] Iterate over `itaac_ids` directly in lieu of a range of numbers to slice into `itaac_ids`.  When you need the index value in addition to the iterator use `enumerate`.

### Refactored `generate_itaac_efforts` method.
The following code incorporates these changes.

```suggestion
    def generate_itaac_efforts(self, itaac_ids):
        """Generate ITAAC Efforts

        :param itaac_ids: IDs of ITAACS to generate efforts
        """
        effort_types = [
            "ITAAC",
            "Program",
            "Reactive/Allegations",
            "Technical",
            "Total",
        ]
        output = []

        with open(f'{self.directory}itaac_efforts.csv', 'w+') as output_file:

            for idx, itaac_id in enumerate(itaac_ids, 1):
                actual = 0
                estimated = 0

                for effort_type in effort_types:
                    if effort_type == 'Total':
                        output.append(f'\n{idx}|'
                                      f'{itaac_id}|'
                                      f'{effort_type}|'
                                      f'{actual}|'
                                      f'{estimated}')
                    else:
                        actual += random.randint(0, 120)
                        estimated += random.randint(0, 120)

            output_file.write("id|itaac_id|effort_type|actual|estimate")
            output_file.write(''.join(output)
```

File: generator/vogtle_data_generator.py
Comment (TimothyHelton) [inline generator/vogtle_data_generator.py]: - [x] Convert to multiline for future header expansion.
- [ ] The old string ended with a blank line which was removed.  Do you still want a new line?

```suggestion
        header = (
            'id|'
            'itaac_status|'
            'icn_status|'
            'est_completion_date|'
            'date_received|'
            'facility|'
            'targeted_flag'
        )
```

File: generator/vogtle_data_generator.py
Comment (TimothyHelton) [inline generator/vogtle_data_generator.py]: - [x] remove `rows`
- [x] add `:param args: ...`
- [x] add `:param config: ...`

**Review by TimothyHelton — CHANGES_REQUESTED** [review]
See task lists in the comments for requested changes.  Please let me know of any of the requests that are not clear and I'll try to explain further.

File: generator/vogtle_data_generator.py
Comment (calderonmluis) [reply generator/vogtle_data_generator.py]: I got pushback from 18F to implement the comments with `%s`, they would not accept my PR until I did so. The original implementation was with the fstrings

**Comment by calderonmluis** [comment]
@TimothyHelton I have made the majority of the requested changes. Please take a look when you get a chance.

File: generator/vogtle_data_generator.py
Comment (TimothyHelton) [reply generator/vogtle_data_generator.py]: Would you please tell me the username of the person who provided you instruction to not use fstrings?

Comment (calderonmluis) [reply generator/vogtle_data_generator.py]: https://github.com/18F/NRC-MAP/pull/1#discussion_r346539788, kfoley-18F

File: generator/vogtle_data_generator.py
Comment (calderonmluis) [reply generator/vogtle_data_generator.py]: resolved

File: generator/vogtle_data_generator.py
Comment (calderonmluis) [reply generator/vogtle_data_generator.py]: resolved

File: generator/vogtle_data_generator.py
Comment (calderonmluis) [reply generator/vogtle_data_generator.py]: resolved

File: generator/vogtle_data_generator.py
Comment (TimothyHelton) [reply generator/vogtle_data_generator.py]: I will speak with kfoley-18f.

[Formatted String Literals](https://docs.python.org/3/tutorial/inputoutput.html?highlight=literal%20string#tut-f-strings) are different than the [string format method()](https://docs.python.org/3/tutorial/inputoutput.html?highlight=string%20format%20method#the-string-format-method) kfoley-18f recommended you change.  

They were introduced in version 3.6 through [PEP 498](https://www.python.org/dev/peps/pep-0498/) and are the best option for string formatting among Python developers with the caveat that they should not be used in logging statements.

I understand that you wish to merge your PR so just leave the old-style string formatting and I'll change all at a later time.

Comment (calderonmluis) [reply generator/vogtle_data_generator.py]: Yes, I know of the difference. There were some ad-hoc chats on the topic.

I will await your approval before merging.

File: generator/vogtle_data_generator.py
Comment (TimothyHelton) [reply generator/vogtle_data_generator.py]: @calderonmluis Did you want a new line on the header?

File: generator/vogtle_data_generator.py
Comment (TimothyHelton) [reply generator/vogtle_data_generator.py:84]: @calderonmluis Would you please document why you chose not to implement the changes listed above?

**Comment by calderonmluis** [comment]
@TimothyHelton I mistakenly had hit the button yesterday and those recommended changes disappeared. I've gone ahead and applied them

**Comment by TimothyHelton** [comment]
@calderonmluis Were you planning on rebasing this branch to the front of develop prior to merging?

File: generator/vogtle_data_generator.py
Comment (TimothyHelton) [inline generator/vogtle_data_generator.py]: @calderonmluis The github "Suggestions" feature cannot be applied to multiline comments, so I put it on the first line of the function.  When you accepted the suggestion only the first line of the previous code was replaced.  Would you please remove lines 114-158.  Sorry about that.

Comment (calderonmluis) [reply generator/vogtle_data_generator.py]: @TimothyHelton I noticed and am working on the fix locally. Also, should I rebase? I can just pull in the changes from `develop`, is there any reason you prefer a rebase instead?

Comment (TimothyHelton) [reply generator/vogtle_data_generator.py]: @calderonmluis I think the only main difference between a rebase and a merge, in this case, is the history.  With a rebase it is very clear where a branch was on the trunk and what commits preceded it.

The tree will look like this:

Screen Shot 2019-12-31 at 8 09 51 AM


I normally rebase a feature branch before submitting a PR and after incorporating any comments from reviewers.  If you rebase during a review all the code shows up as "changed" making it hard on the reviewers.

```bash
git pull --rebase origin develop
```

A rebase before submitting the PR lets me see any merge conflicts that need to be addressed.

A rebase after the review ensures another PR wasn't merged that causes me trouble.  This is also a great time to consolidate fixup commits that were made for reviewers.

### An example where O represents a commit:
O ENH: new function 1
O ENH: new function 2
git pull --rebase origin develop
O FIX: update function 1
Rebase FIX into ENH: new function 1
Submit PR
O FIX: address xyz issue for function 1 from reviewer 3
O FIX: address mno issue for function 2 from reviewer 5
Review is complete
git pull --rebase origin develop
Rebase FIX commits into original ENH commits
Merge PR

**Review by TimothyHelton — APPROVED** [review]

File: generator/vogtle_data_generator.py
Comment (calderonmluis) [reply generator/vogtle_data_generator.py]: I went ahead and ran the rebase. Thanks for the approval

**Pull request merged**
